libraries/tlv-account-resolution/src/key_data.rs

```diff
@@ -0,0 +1,183 @@
+
+/// Enum to describe a required key stored in some data.
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "serde-traits", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde-traits", serde(rename_all = "camelCase"))]
+pub enum KeyData {
+    /// Uninitialized configuration byte space.
+    Uninitialized,
+    /// A key to be resolved from the instruction data.
+    ///
+    /// Packed as:
+    ///     * 1 - Discriminator
+    ///     * 1 - Start index of instruction data
+    ///
+    /// Note: Length is always 32 bytes.
+    InstructionData {
+        /// The index where the address bytes begin in the instruction data.
+        index: u8,
+    },
+    /// A key to be resolved from the inner data of some account.
+    ///
+    /// Packed as:
+    ///     * 1 - Discriminator
+    ///     * 1 - Index of account in accounts list
+    ///     * 1 - Start index of account data
+    ///
+    /// Note: Length is always 32 bytes.
+    AccountData {
+        /// The index of the account in the entire accounts list.
+        account_index: u8,
+        /// The index where the address bytes begin in the account data.
+        data_index: u8,
+    },
+}
+impl KeyData {
+    /// Get the size of a key data configuration.
+    pub fn tlv_size(&self) -> u8 {
+        match self {
+            Self::Uninitialized => 0,
+            // 1 byte for the discriminator, 1 byte for the index.
+            Self::InstructionData { .. } => 1 + 1,
+            // 1 byte for the discriminator, 1 byte for the account index,
+            // 1 byte for the data index.
+            Self::AccountData { .. } => 1 + 1 + 1,
+        }
+    }
+
+    /// Packs a key data configuration into a slice.
+    pub fn pack(&self, dst: &mut [u8]) -> Result<(), ProgramError> {
+        // Because no `KeyData`` variant is larger than 3 bytes, this check is
+        // sufficient for the data length.
+        if dst.len() != self.tlv_size() as usize {
+            return Err(AccountResolutionError::NotEnoughBytesForKeyData.into());
+        }
+        match &self {
+            Self::Uninitialized => return Err(AccountResolutionError::InvalidKeyDataConfig.into()),
+            Self::InstructionData { index } => {
+                dst[0] = 1;
+                dst[1] = *index;
+            }
+            Self::AccountData {
+                account_index,
+                data_index,
+            } => {
+                dst[0] = 2;
+                dst[1] = *account_index;
+                dst[2] = *data_index;
+            }
+        }
+        Ok(())
+    }
+
+    /// Packs a key data configuration into a 32-byte array, filling the rest
+    /// with 0s.
+    pub fn pack_into_address_config(key_data: &Self) -> Result<[u8; 32], ProgramError> {
+        let mut packed = [0u8; 32];
+        let tlv_size = key_data.tlv_size() as usize;
+        key_data.pack(&mut packed[..tlv_size])?;
+        Ok(packed)
+    }
```

Comment:
It's nice when the more complicated variant is already done, and this can be simple :sweat_smile:

Reply:
For real! We didn't even need `unpack_address_config` for this one.

Reply:
The most beautiful part was not having to change anything in `ExtraAccountMetaList`. 😎